substrate/client/transaction-pool/src/fork_aware_txpool/dropped_watcher.rs

```diff
@@ -279,12 +286,23 @@ where
 						return Some(DroppedTransaction::new_enforced_by_limts(tx_hash))
 					}
 				} else {
-					debug!(target: LOG_TARGET, ?tx_hash, "dropped_watcher: removing (non-tracked) tx");
+					debug!(target: LOG_TARGET, ?tx_hash, "dropped_watcher: removing (non-tracked dropped) tx");
 					return Some(DroppedTransaction::new_enforced_by_limts(tx_hash))
 				}
 			},
 			TransactionStatus::Usurped(by) =>
 				return Some(DroppedTransaction::new_usurped(tx_hash, by)),
+			TransactionStatus::Invalid => {
+				if let Some(mut views_keeping_tx_valid) = self.transaction_views(tx_hash) {
```

Comment:
Hmm, `transaction_views` only returns either the future or ready views. But in this case, should the block hash not get removed from both?

For reference:
https://github.com/paritytech/polkadot-sdk/blob/0910ead7b0331bf25cb66107dc5ae5a1f8ec5f08/substrate/client/transaction-pool/src/fork_aware_txpool/dropped_watcher.rs#L174-L189

Reply:
Good point. 

`transaction_views` returns entry pointing to `ready_transaction_views` or `future_transaction_views` map.

 It is also not possible that transaction is in both future and ready, see here:
https://github.com/paritytech/polkadot-sdk/blob/0910ead7b0331bf25cb66107dc5ae5a1f8ec5f08/substrate/client/transaction-pool/src/fork_aware_txpool/dropped_watcher.rs#L264-L281

(I fixed flow for future).

Reply:
Added some extra check in: 6d7ccff